File: data_structures/queue/queue.py
class Queue:
  class LinkedList:
    class Node:
      def __init__(self):
        self.value = None
        self.prev = None
        self.next = None

    def __init__(self):
      self._size = 0
      self.head = self.Node()
      self.tail = self.Node()

    def push_front(self, value):
      node = self.Node()
      node.value = value
      if self._size == 0:
        self.head.next = node
        self.tail.prev = node
      else:
        node.next = self.head.next
        self.head.next.prev = node
        self.head.next = node
      self._size+=1

    def pop_back(self):
      if self._size == 0:
        return -1
      else:
        value = self.tail.prev.value

        if self._size == 1:
          self.tail.prev = None
          self.head.next = None
        else:
          self.tail.prev = self.tail.prev.prev
        self._size-=1
        return value
  def __init__(self):
    self.list = self.LinkedList()

  def enqueue(self, value):
    self.list.push_front(value)
  
  def dequeue(self):
    return self.list.pop_back()

  def empty(self):
    return self.list._size == 0

q = Queue()

q.enqueue(10)
q.enqueue(20)
q.enqueue(30)
q.enqueue(40)

print(q.dequeue())
print(q.dequeue())
print(q.dequeue())
print(q.dequeue())
print(q.dequeue())

File: data_structures/queue/queue.cpp
#include <stdio.h>
#include <stdlib.h>

class LinkedList {
  struct Node {
    int value;
    Node* next;
    Node* prev;
  };

  Node* head;
  Node* tail;
  int size;

  public:

  LinkedList() {
    head = new Node();
    tail = new Node();
    size = 0;
  }

  void push_front(int value) {
    Node* node = new Node();
    node->value = value;

    if (size == 0) {
      head->next = node;
      tail->prev = node;
    }
    else {
      head->next->prev = node;
      node->next = head->next;
      head->next = node;
    }
    size++;
  }

  int pop_back() {
    if (size == 0) {
      return -1;
    }
    else {
      int value = tail->prev->value;
      Node* node = tail->prev;
      
      if (size == 1) {
        tail->prev = nullptr;
        head->next = nullptr;
      }
      else {
        tail->prev = tail->prev->prev;
      }

      delete node;
      size--;
      return value;
    }
  }

  int front() {
    if (size == 0) {
      return -1;
    }
    else {
      return head->next->value;
    }
  }

  int back() {
    if (size == 0) {
      return -1;
    }
    else {
      return tail->prev->value;
    }
  }
};

int main() {
  LinkedList list;

  list.push_front(10);
  list.push_front(20);
  list.push_front(30);
  list.push_front(40);

  printf("Front %d\n", list.front());
  printf("Back %d\n\n", list.back());

  printf("Finishing job %d\n\n", list.pop_back());
  printf("Finishing job %d\n\n", list.pop_back());
  printf("Finishing job %d\n\n", list.pop_back());
  printf("Finishing job %d\n\n", list.pop_back());
}